README.md
# Letsconsul

**STATUS:** *prototyping/experimental*

Tool that helps you automatically renew LetsEncrypt certificates and serve them in consul K/V storage.

*Note: It is tightly integrated with proxy-server (nginx in this case) and `consul-template` tool. Please read below for full understanding of certificate issuing and installation process.*

## Get started

At first create following K/V structure in Consul:

```
letsconsul
|_
| \_renew_interval = 168h
|_
| \_reload_interval = 10s
|_
| \_service = letsconsul
|_
| \_domains_enabled = ["example.com", "qlean.ru"]
|_
   \_domains
     |_
     | \_example.com
     |   |_
     |   | \_domain_list = ["www.example.com", "example.com"]
     |   |_
     |   | \_email = admin@example.com
     |   |_
     |   | \_timestamp = 0
     |   |_
     |   | \_private_key =
     |   |_
     |     \_fullchain =
     |_
       \_qlean.ru
         |_
         | \_domain_list = ["qlean.ru", "www.qlean.ru", "assets.qlean.ru"]
         |_
         | \email = webmaster@qlean.ru
         |_
         | \_timestamp = 0
         |_
         | \_private_key =
         |_
           \_fullchain =
```

Consul configuration keys:

- `renew_interval` - domain certificate expiration time
- `reload_interval` - time after letsconsul reloading domains information from consul
- `service` - consul service name
- `domains_enabled` - domains from `letsconsul/domains` that can be validated and renewed with certs/keys

When letsconsul starting it reading particular command line arguments:

- `-b` - host:port variable that validation web-server will listen (by default 0.0.0.0:8080)
- `-c` - consul address (by default 127.0.0.1:8500)

Also, you can specifly consul ACL token with CONSUL_TOKEN environment variable.

Example of usage:

```
$ wget https://github.com/hypersleep/letsconsul/releases/download/0.0.2/letsconsul-linux-64.zip
$ unzip letsconsul-linux-64.zip
$ ./letsconsul -b 0.0.0.0:8080 -c 127.0.0.1:8500
```

## Workflow description

After app starts, it fetching domains information from consul by given `consul_service` key, checking certificate expiration time and if more than `renew_interval` then starts certificate renew process.

Updating and receiving certificates is based on LetsEncrypt HTTP validation.

After validation request has sent to LetsEncrypt, letsconsul starts a validation web-server on address `-b` that should receive a LetsEncrypt validation request.

Ensure that validation web-server available from internet! You can simply do this with combination of nginx proxy and consul-template:

```
server {
    listen 80;
    server_name www.example.com example.com;

    {{range service "letsconsul"}}
    location /.well-known/acme-challenge {
        proxy_set_header Host $host;
        proxy_pass http://{{.Address}}:{{.Port}};
    }
    {{end}}
}
```

If validation will be made successfully, letsconsul writes received certificates and keys to `fullchain` and `private_key` consul keys.

After that you can use values of `fullchain` and `private_key` as cert/key files for nginx using consul-template:

/etc/ssl/example.com.crt.ctmpl:
```
{{key "letsconsul/domains/example.com/fullchain"}}
```
rendering to: /etc/ssl/example.com.crt

/etc/ssl/example.com.key.ctmpl:
```
{{key "letsconsul/domains/example.com/private_key"}}
```
rendering to: /etc/ssl/example.com.key

Finally, you're got fully automated and distributed by many servers/proxies HTTPS certificates!

You can see full workflow on following chart:

![Workflow](workflow.png)


CHANGELOG.md
## 0.0.2

- errors handling #3
- complete readme #2
- better configuration #1


domains.go
package main

import(
	"log"
	"time"
	"errors"
	"strconv"
	// "bytes"
	"encoding/json"

	// "rsc.io/letsencrypt"
	consul "github.com/hashicorp/consul/api"
)

type(
	Letsconsul struct {
		Domains map[string]*DomainRecord
	}

	DomainRecord struct {
		Domains   []string  `consul:"domains"`
		Timestamp time.Time `consul:"timestamp"`
		Cert      string    `consul:"cert"`
		Chain     string    `consul:"chain"`
		Fullchain string    `consul:"fullchain"`
	}
)

func (domainRecord *DomainRecord) renew(name string) error {
	//return  errors.New("Failed to renew domain record: " + name +" Reason: " + err)
	log.Println(domainRecord, name)
	return nil
}

func (letsconsul *Letsconsul) renew(renewInterval time.Duration) error {
	// for name, domainRecord := range letsconsul.Domains {
	// 	timestamp, err := time.Parse(time.RFC3339, domainRecord.Timestamp)
	// 	if err != nil {
	// 		return err
	// 	}

	// 	err = domainRecord.renew(name)
	// 	if err != nil {
	// 		return err
	// 	}
	// }
	return nil
}

func kvFetch(kv *consul.KV, prefix string, domainRecordName string, key string) ([]byte, error) {
	kvPair, _, err := kv.Get(prefix + "/domains/" + domainRecordName + "/" + key, nil)
	if err != nil {
		return nil, err
	}

	if kvPair == nil {
		return nil, errors.New("Can't fetch '" + key + "' key from '" + domainRecordName + "' domain")
	}

	return kvPair.Value, nil
}

func (domainRecord *DomainRecord) get(kv *consul.KV, prefix string, domainRecordName string) error {
	v, err := kvFetch(kv, prefix, domainRecordName, "domain_list")
	if err != nil {
		return err
	}

	err = json.Unmarshal(v, &domainRecord.Domains)
	if err != nil {
		return err
	}


	v, err = kvFetch(kv, prefix, domainRecordName, "timestamp")
	if err != nil {
		return err
	}

	i, err := strconv.ParseInt(string(v), 10, 64)
	if err != nil {
		return err
	}

	domainRecord.Timestamp = time.Unix(i, 0)

	v, err = kvFetch(kv, prefix, domainRecordName, "cert")
	if err != nil {
		return err
	}

	domainRecord.Cert = string(v)

	v, err = kvFetch(kv, prefix, domainRecordName, "chain")
	if err != nil {
		return err
	}

	domainRecord.Chain = string(v)

	v, err = kvFetch(kv, prefix, domainRecordName, "fullchain")
	if err != nil {
		return err
	}

	domainRecord.Fullchain = string(v)


	return nil
}

func (letsconsul *Letsconsul) get(client *consul.Client, prefix string) error {
	kv := client.KV()

	kvPair, _, err := kv.Get(prefix + "/domains_enabled", nil)
	if err != nil {
		return err
	}

	if kvPair == nil {
		return errors.New("Can't fetch 'domains_enabled' key")
	}

	var domainsEnabled []string = []string{}

	err = json.Unmarshal(kvPair.Value, &domainsEnabled)
	if err != nil {
		return err
	}

	for i := range domainsEnabled {
		if letsconsul.Domains[domainsEnabled[i]] == nil {
			domainRecord := &DomainRecord{}
			err := domainRecord.get(kv, prefix, domainsEnabled[i])
			if err != nil {
				return err
			}
			letsconsul.Domains[domainsEnabled[i]] = domainRecord
		}
	}

	return nil
}


letsconsul.go
package main

import(
	"time"
	"errors"
	"encoding/json"

	consul "github.com/hashicorp/consul/api"
)

type(
	Letsconsul struct {
		Bind    string
		Domains map[string]*DomainRecord
	}
)

func (letsconsul *Letsconsul) renew(client *consul.Client, renewInterval time.Duration) error {
	for domainRecordName, domainRecord := range letsconsul.Domains {
		if domainRecord.Timestamp.Add(renewInterval).Before(time.Now()) {
			err := domainRecord.renew(letsconsul.Bind)
			if err != nil {
				return err
			}

			err = domainRecord.write(client, domainRecordName)
			if err != nil {
				return err
			}
		}
	}
	return nil
}

func (letsconsul *Letsconsul) get(client *consul.Client) error {
	kv := client.KV()
	prefix := "letsconsul"

	kvPair, _, err := kv.Get(prefix + "/domains_enabled", nil)
	if err != nil {
		return err
	}

	if kvPair == nil {
		return errors.New("Can't fetch 'domains_enabled' key")
	}

	var domainsEnabled []string = []string{}

	err = json.Unmarshal(kvPair.Value, &domainsEnabled)
	if err != nil {
		return err
	}

	for i := range domainsEnabled {
		if letsconsul.Domains[domainsEnabled[i]] == nil {
			domainRecord := &DomainRecord{}
			err := domainRecord.get(kv, prefix, domainsEnabled[i])
			if err != nil {
				return err
			}
			letsconsul.Domains[domainsEnabled[i]] = domainRecord
		}
	}
	return nil
}


domain_record.go
package main

import(
	"fmt"
	"net"
	"time"
	"crypto"
	"crypto/elliptic"
	"crypto/rand"
	"crypto/ecdsa"
	"errors"
	"strconv"
	"encoding/json"

	"github.com/xenolf/lego/acme"
	consul "github.com/hashicorp/consul/api"
)

type(
	DomainRecord struct {
		Domains    []string  `consul:"domains"`
		Email      string    `consul:"email"`
		Timestamp  time.Time `consul:"timestamp"`
		PrivateKey string    `consul:"private_key"`
		Fullchain  string    `consul:"fullchain"`
		key   *ecdsa.PrivateKey
		Reg   *acme.RegistrationResource
	}
)

func (domainRecord *DomainRecord) GetEmail() string { return domainRecord.Email }
func (domainRecord *DomainRecord) GetRegistration() *acme.RegistrationResource { return domainRecord.Reg }
func (domainRecord *DomainRecord) GetPrivateKey() crypto.PrivateKey { return domainRecord.key }

func (domainRecord *DomainRecord) write(client *consul.Client, domainRecordName string) error {
	kv := client.KV()
	consulService := "letsconsul"

	timestamp := domainRecord.Timestamp.Unix()
	timestampStr := strconv.Itoa(int(timestamp))

	p := &consul.KVPair {
		Key: consulService + "/domains/" + domainRecordName + "/timestamp",
		Value: []byte(timestampStr),
	}
	_, err := kv.Put(p, nil)
	if err != nil {
		return err
	}

	p = &consul.KVPair {
		Key: consulService + "/domains/" + domainRecordName + "/fullchain",
		Value: []byte(domainRecord.Fullchain),
	}
	_, err = kv.Put(p, nil)
	if err != nil {
		return err
	}

	p = &consul.KVPair {
		Key: consulService + "/domains/" + domainRecordName + "/private_key",
		Value: []byte(domainRecord.PrivateKey),
	}
	_, err = kv.Put(p, nil)
	if err != nil {
		return err
	}

	return nil
}

func (domainRecord *DomainRecord) renew(bind string) error {
	const letsEncryptURL = "https://acme-v01.api.letsencrypt.org/directory"

	key, err := ecdsa.GenerateKey(elliptic.P384(), rand.Reader)
	if err != nil {
		return err
	}

	domainRecord.key = key

	client, err := acme.NewClient(letsEncryptURL, domainRecord, acme.EC256)
	if err != nil {
		return err
	}

	reg, err := client.Register()
	if err != nil {
		return err
	}

	domainRecord.Reg = reg

	err = client.AgreeToTOS()
	if err != nil {
		return err
	}

	host, port, err := net.SplitHostPort(bind)
	if err != nil {
		return err
	}

	client.ExcludeChallenges([]acme.Challenge{acme.TLSSNI01})
	client.ExcludeChallenges([]acme.Challenge{acme.HTTP01})

	httpProvider := acme.NewHTTPProviderServer(host, port)

	err = client.SetChallengeProvider(acme.HTTP01, httpProvider)
	if err != nil {
		return err
	}

	acmeCert, errmap := client.ObtainCertificate(domainRecord.Domains, true, nil)
	if len(errmap) > 0 {
		err = fmt.Errorf("%v", errmap)
		return err
	}

	domainRecord.Fullchain = string(acmeCert.Certificate)
	domainRecord.PrivateKey = string(acmeCert.PrivateKey)

	domainRecord.Timestamp = time.Now()

	return nil
}

func kvFetch(kv *consul.KV, prefix string, domainRecordName string, key string) ([]byte, error) {
	kvPair, _, err := kv.Get(prefix + "/domains/" + domainRecordName + "/" + key, nil)
	if err != nil {
		return nil, err
	}

	if kvPair == nil {
		return nil, errors.New("Can't fetch '" + key + "' key from '" + domainRecordName + "' domain")
	}

	return kvPair.Value, nil
}

func (domainRecord *DomainRecord) get(kv *consul.KV, prefix string, domainRecordName string) error {
	v, err := kvFetch(kv, prefix, domainRecordName, "domain_list")
	if err != nil {
		return err
	}

	err = json.Unmarshal(v, &domainRecord.Domains)
	if err != nil {
		return err
	}

	v, err = kvFetch(kv, prefix, domainRecordName, "email")
	if err != nil {
		return err
	}

	domainRecord.Email = string(v)

	v, err = kvFetch(kv, prefix, domainRecordName, "timestamp")
	if err != nil {
		return err
	}

	i, err := strconv.ParseInt(string(v), 10, 64)
	if err != nil {
		return err
	}

	domainRecord.Timestamp = time.Unix(i, 0)

	return nil
}


app.go
package main

import(
	"os"
	"log"
	"net"
	"time"
	"flag"
	"errors"
	"net/http"
	"strconv"
	"syscall"
	"os/signal"

	"github.com/satori/go.uuid"
	consul "github.com/hashicorp/consul/api"
)

type(
	App struct {
		Bind            string
		ConsulToken     string
		ConsulService   string        `consul:"service"`
		RenewInterval   time.Duration `consul:"renew_interval"`
		ReloadInterval  time.Duration `consul:"reload_interval"`
		consulClient    *consul.Client
		consulServiceID string
		letsconsul      *Letsconsul
	}
)

func (app *App) consulConfigure() error {
	kv := app.consulClient.KV()
	prefix := "letsconsul"

	kvPair, _, err := kv.Get(prefix + "/service", nil)
	if err != nil {
		return err
	}

	if kvPair == nil {
		return errors.New("Can't fetch 'service' key")
	}

	app.ConsulService = string(kvPair.Value)

	kvPair, _, err = kv.Get(prefix + "/renew_interval", nil)
	if err != nil {
		return err
	}

	if kvPair == nil {
		return errors.New("Can't fetch 'renew_interval' key")
	}

	app.RenewInterval, err = time.ParseDuration(string(kvPair.Value))
	if err != nil {
		return err
	}

	kvPair, _, err = kv.Get(prefix + "/reload_interval", nil)
	if err != nil {
		return err
	}

	if kvPair == nil {
		return errors.New("Can't fetch 'reload_interval' key")
	}

	app.ReloadInterval, err = time.ParseDuration(string(kvPair.Value))
	if err != nil {
		return err
	}

	return nil
}

func (app *App) config() error {
	app.ConsulToken = os.Getenv("CONSUL_TOKEN")

	bindPtr := flag.String("b", "0.0.0.0:8080", "host:port variable that validation web-server will listen")
	consulAddrPtr := flag.String("c", "127.0.0.1:8500", "consul address")
	flag.Parse()

	app.Bind = *bindPtr

	consulConfig := &consul.Config{
		Address:    *consulAddrPtr,
		Token:      app.ConsulToken,
		Scheme:     "http",
		HttpClient: http.DefaultClient,
	}

	client, err := consul.NewClient(consulConfig)
	if err != nil {
		return err
	}

	app.consulClient = client

	err = app.consulConfigure()
	if err != nil {
		return err
	}

	go func() {
		for {
			time.Sleep(10 * time.Second)
			app.consulConfigure()
		}
	}()

	return nil
}

func (app *App) register() error {
	app.consulServiceID = uuid.NewV4().String()

	checks := consul.AgentServiceChecks{
		&consul.AgentServiceCheck{
			TTL: "5s",
		},
	}

	_, portstr, err := net.SplitHostPort(app.Bind)
	if err != nil {
		return err
	}

	port, err := strconv.Atoi(portstr)
	if err != nil {
		return err
	}

	service := &consul.AgentServiceRegistration{
		ID: app.consulServiceID,
		Name: app.ConsulService,
		Port: port,
		Checks: checks,
	}

	agent := app.consulClient.Agent()
	err = agent.ServiceRegister(service)
	if err != nil {
		return err
	}

	go func() {
		for {
			<- time.After(2 * time.Second)
			err = agent.PassTTL("service:" + app.consulServiceID, "Internal TTL ping")
			if err != nil {
				log.Println(err)
			}
		}
	}()

	go func() {
		signalCh := make(chan os.Signal, 4)
		signal.Notify(signalCh, os.Interrupt, syscall.SIGTERM)
		<-signalCh
		log.Println("Interruption signal caught, exiting...")
		log.Println("Deregister service with ID:", app.consulServiceID)
		err := agent.ServiceDeregister(app.consulServiceID)
		if err != nil {
			log.Fatal("Deregistration failed. Reason:", err)
		} else {
			log.Println("Deregistration successful. Exiting with 0 exit code.")
			os.Exit(0)
		}
	}()

	time.Sleep(5 * time.Second)

	return nil
}

func (app *App) renewDomains() error {
	app.letsconsul.Domains = make(map[string]*DomainRecord)

	err := app.letsconsul.get(app.consulClient)
	if err != nil {
		return err
	}

	app.letsconsul.Bind = app.Bind

	err = app.letsconsul.renew(app.consulClient, app.RenewInterval)
	if err != nil {
		return err
	}

	return nil
}

func (app *App) start() {
	app.letsconsul = &Letsconsul{}

	for {
		err := app.renewDomains()
		if err != nil {
			log.Println(err)
		}

		<- time.After(app.ReloadInterval)
	}
}
